### PR TITLE
Implement function sampling an index weighted by probability

### PR DESCRIPTION
**Pull request by Benzbeeb**

### CODE REVIEW

File: chain/pkg/bandrng/sampling.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package bandrng
+
+import "math"
+
+// AddUint64Overflow performs the addition operation on two uint64 integers and
+// returns a boolean on whether or not the result overflows.
+func AddUint64Overflow(a, b uint64) (uint64, bool) {
```
Comment (sorawit) [inline chain/pkg/bandrng/sampling.go]: I think this should be private.

File: chain/pkg/bandrng/sampling.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	if math.MaxUint64-a < b {
+		return 0, true
+	}
+
+	return a + b, false
+}
+
+// SamplingOne sampling an index weighted by probability
+func SamplingOne(rng *Rng, weights []uint64) int {
```
Comment (sorawit) [inline chain/pkg/bandrng/sampling.go]: `// ChooseOne randomly picks an index between 0 and len(weights)-1 inclusively. Each index has the probability of getting selected based on its weight.`

File: chain/pkg/bandrng/sampling.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	sum := uint64(0)
+	var overflow bool
+	for _, weight := range weights {
+		sum, overflow = AddUint64Overflow(sum, weight)
+		if overflow {
+			panic("a sum of weight is more than max uint64")
```
Comment (sorawit) [inline chain/pkg/bandrng/sampling.go]: `panic("sum of weights exceed max uint64")`

File: chain/pkg/bandrng/sampling.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		}
+	}
+
+	luckyNumber := rng.NextUint64() % sum
+	count := uint64(0)
+	for idx, weight := range weights {
```
Comment (sorawit) [inline chain/pkg/bandrng/sampling.go:29]: Can just +=. There's no way this overflows.

File: chain/pkg/bandrng/sampling.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	var overflow bool
+	for _, weight := range weights {
+		sum, overflow = AddUint64Overflow(sum, weight)
+		if overflow {
+			panic("a sum of weight is more than max uint64")
+		}
+	}
+
+	luckyNumber := rng.NextUint64() % sum
+	count := uint64(0)
```
Comment (sorawit) [inline chain/pkg/bandrng/sampling.go]: currentSum

File: chain/pkg/bandrng/sampling.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		count, overflow = AddUint64Overflow(count, weight)
+		if overflow {
+			panic("a um of weight is more than max uint64")
+
+		}
+		if count > luckyNumber {
+			return idx
+		}
+	}
+	// Should never happen because the sum of weight is more than a lucky number
```
Comment (sorawit) [inline chain/pkg/bandrng/sampling.go]: weights ... the lucky number

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**